play-json/src/main/scala/play/api/libs/json/JsMacroImpl.scala

```diff
@@ -256,227 +363,273 @@ object JsMacroImpl {
+        case Some(_) => allSubclasses(path.tail, subclasses)
 
-        case _ => ptype
+        case _ => subclasses
       }
 
-      val forwardName = TermName(c.freshName("forward"))
+      if (tpeSym.isSealed && tpeSym.isAbstract) {
+        Some(allSubclasses(tpeSym.owner.typeSignature.decls, Set.empty).toList)
+      } else None
+    }
 
-      private object ImplicitTransformer extends Transformer {
-        override def transform(tree: Tree): Tree = tree match {
-          case tt: TypeTree =>
-            super.transform(TypeTree(denormalized(tt.tpe)))
+    // --- Sub implementations
 
-          case Select(Select(This(TypeName("JsMacroImpl")), t), sym) if (
-            t.toString == "Placeholder" && sym.toString == "Format"
-          ) => super.transform(q"$forwardName")
+    val readsType = c.typeOf[Reads[_]]
 
-          case _ => super.transform(tree)
-        }
-      }
+    def macroSealedFamilyImpl(subTypes: List[Type]): c.Expr[M[A]] = {
+      val typeNaming = (_: Type).typeSymbol.fullName
 
-      private def createImplicit(name: Name, ptype: Type): Implicit = {
-        val tpe = ptype match {
-          case TypeRef(_, _, targ :: _) =>
-            // Option[_] needs special treatment because we need to use XXXOpt
-            if (ptype.typeConstructor <:< typeOf[Option[_]].typeConstructor) {
-              targ
-            } else ptype
+      def readLambda: Tree = {
+        val resolver = new ImplicitResolver({ orig: Type => orig })
+        val cases = Match(q"""obj.value.lift("_value")""", (
+          cq"""_ => $json.JsError("error.missing.path")""" :: (
+            subTypes.foldLeft(List.empty[CaseDef]) { (out, t) =>
+              val rtpe = appliedType(readsType, List(t))
+              val reader = resolver.createImplicit(
+                atag.tpe, rtpe)(t).neededImplicit
 
-          case SingleType(_, _) | TypeRef(_, _, _) => ptype
-        }
+              if (reader.isEmpty) {
+                c.abort(c.enclosingPosition,
+                  s"No instance of Reads is available for ${t.typeSymbol.fullName} in the implicit scope (Hint: if declared in the same file, make sure it's declared before)")
+              }
 
-        val (ntpe, selfRef) = normalized(tpe)
-        val ptpe = boundTypes.get(ntpe.typeSymbol.fullName).getOrElse(ntpe)
+              cq"""Some(v) if (dis == ${typeNaming(t)}) => $reader.reads(v)""" :: out
+            }
+          )
+        ).reverse)
 
-        // infers implicit
-        val neededImplicitType = appliedType(natag.tpe.typeConstructor, ptpe)
-        val neededImplicit = if (!selfRef) {
-          c.inferImplicitValue(neededImplicitType)
-        } else c.untypecheck(
-          // Reset the type attributes on the refactored tree for the implicit
-          ImplicitTransformer.transform(
-            c.inferImplicitValue(neededImplicitType))
-        )
+        q"""(_: $json.JsValue) match {
+          case obj @ $json.JsObject(_) => obj.value.get("_type") match {
```

Comment:
Yes, but not sure it should be done right in this PR, as it may be related with some refactoring of `JsonConfiguration`